File: firmware/src/osc_sender.cpp
/**
 * VRC Facial Tracker - OSC Sender Module
 */

#include "osc_sender.h"
#include "config.h"
#include <WiFiUdp.h>
#include <OSCMessage.h>

static WiFiUDP udp;
static OscStats stats = {};

void osc_init() {
    udp.begin(OSC_LOCAL_PORT);
    Serial.printf("OSC: Sending to %s:%d\n", OSC_TARGET_IP, OSC_TARGET_PORT);
}

/**
 * Send a single float OSC parameter.
 */
static bool send_param(const char *address, float value) {
    OSCMessage msg(address);
    msg.add(value);
    udp.beginPacket(OSC_TARGET_IP, OSC_TARGET_PORT);
    msg.send(udp);
    bool ok = udp.endPacket() == 1;
    msg.empty();
    return ok;
}

void osc_send(const FaceData &face) {
    // VRChat face tracking OSC addresses
    // Reference: https://docs.vrchat.com/docs/osc-as-input-controller

    // Eyes
    bool ok = true;

    // Eyes
    ok &= send_param("/avatar/parameters/EyeClosedLeft",    face.eyeClosedLeft);
    ok &= send_param("/avatar/parameters/EyeClosedRight",   face.eyeClosedRight);
    ok &= send_param("/avatar/parameters/EyeSquintLeft",     face.eyeSquintLeft);
    ok &= send_param("/avatar/parameters/EyeSquintRight",    face.eyeSquintRight);
    ok &= send_param("/avatar/parameters/EyeWideLeft",       face.eyeWideLeft);
    ok &= send_param("/avatar/parameters/EyeWideRight",      face.eyeWideRight);

    // Eyebrows
    ok &= send_param("/avatar/parameters/BrowUpLeft",        face.browUpLeft);
    ok &= send_param("/avatar/parameters/BrowUpRight",       face.browUpRight);
    ok &= send_param("/avatar/parameters/BrowDownLeft",      face.browDownLeft);
    ok &= send_param("/avatar/parameters/BrowDownRight",     face.browDownRight);

    // Mouth
    ok &= send_param("/avatar/parameters/MouthOpen",         face.mouthOpen);
    ok &= send_param("/avatar/parameters/MouthSmile",        face.mouthSmile);
    ok &= send_param("/avatar/parameters/MouthFrown",        face.mouthFrown);
    ok &= send_param("/avatar/parameters/MouthPucker",       face.mouthPucker);

    // Jaw
    ok &= send_param("/avatar/parameters/JawOpen",           face.jawOpen);
    ok &= send_param("/avatar/parameters/JawLeft",           face.jawLeft);
    ok &= send_param("/avatar/parameters/JawRight",          face.jawRight);

    // Cheek
    ok &= send_param("/avatar/parameters/CheekPuff",         face.cheekPuff);

    // Tongue
    ok &= send_param("/avatar/parameters/TongueOut",         face.tongueOut);

    if (ok) {
        stats.total_success++;
        stats.interval_success++;
    } else {
        stats.total_failure++;
        stats.interval_failure++;
    }
}

OscStats osc_get_stats() {
    return stats;
}

void osc_reset_interval_stats() {
    stats.interval_success = 0;
    stats.interval_failure = 0;
}

File: pc_app/src/osc_sender.py
"""
VRC Facial Tracker — OSC sender for VRChat.

Sends face tracking parameters to VRChat via the OSC protocol.
VRChat listens on localhost:9000 for avatar parameters.

Supports two calling conventions:
  - send_blendshapes(dict)  — ARKit blend shape name → float  (used by GUI/CLI)

Reference:
  https://docs.vrchat.com/docs/osc-as-input-controller
"""

from __future__ import annotations

from pythonosc import udp_client

# VRChat OSC address prefix
PREFIX = "/avatar/parameters/"

# MediaPipe ARKit blend shape name → VRChat parameter name
BLENDSHAPE_TO_VRC: dict[str, str] = {
    "eyeBlinkLeft":     "EyeClosedLeft",
    "eyeBlinkRight":    "EyeClosedRight",
    "eyeWideLeft":      "EyeWideLeft",
    "eyeWideRight":     "EyeWideRight",
    "eyeSquintLeft":    "EyeSquintLeft",
    "eyeSquintRight":   "EyeSquintRight",
    # eyeLook* params removed — camera is mounted under VR headset,
    # so iris tracking is impossible.  Blink/wide/squint may still
    # partially work from the lower-face view.
    "browDownLeft":     "BrowDownLeft",
    "browDownRight":    "BrowDownRight",
    "browInnerUp":      "BrowInnerUp",
    "browOuterUpLeft":  "BrowOuterUpLeft",
    "browOuterUpRight": "BrowOuterUpRight",
    "mouthSmileLeft":   "MouthSmileLeft",
    "mouthSmileRight":  "MouthSmileRight",
    "mouthFrownLeft":   "MouthFrownLeft",
    "mouthFrownRight":  "MouthFrownRight",
    "mouthPucker":      "MouthPucker",
    "mouthLeft":        "MouthLeft",
    "mouthRight":       "MouthRight",
    "mouthFunnel":      "MouthFunnel",
    "mouthShrugUpper":  "MouthShrugUpper",
    "mouthShrugLower":  "MouthShrugLower",
    "jawOpen":          "JawOpen",
    "jawLeft":          "JawLeft",
    "jawRight":         "JawRight",
    "jawForward":       "JawForward",
    "cheekPuff":        "CheekPuff",
    "cheekSquintLeft":  "CheekSquintLeft",
    "cheekSquintRight": "CheekSquintRight",
    "tongueOut":        "TongueOut",
    "noseSneerLeft":    "NoseSneerLeft",
    "noseSneerRight":   "NoseSneerRight",
}


class OscSender:
    """Sends facial parameters to VRChat over OSC/UDP."""

    def __init__(self, ip: str = "127.0.0.1", port: int = 9000):
        self._client = udp_client.SimpleUDPClient(ip, port)
        self._sent = 0

    def send_blendshapes(self, blend_shapes: dict[str, float]) -> None:
        """Send ARKit blend shape dict as VRChat OSC messages."""
        for bs_name, vrc_name in BLENDSHAPE_TO_VRC.items():
            val = blend_shapes.get(bs_name, 0.0)
            self._client.send_message(PREFIX + vrc_name, float(val))
        self._sent += 1

    def reconfigure(self, ip: str, port: int) -> None:
        """Change OSC target at runtime (for GUI)."""
        self._client = udp_client.SimpleUDPClient(ip, port)

    @property
    def messages_sent(self) -> int:
        return self._sent
